### PR TITLE
feat(mcp): add unified get_schema tool for schema discovery

### PR DESCRIPTION
**Pull request by aminghadersohi**
### SUMMARY

Consolidates three separate schema discovery tools (`get_chart_available_filters`, `get_dataset_available_filters`, `get_dashboard_available_filters`) into a single unified `get_schema` tool that returns comprehensive metadata for any model type.

#### Token Usage Improvement

**Tool Definition Overhead (loaded into LLM context each session):**
| Metric | Before | After | Savings |
|--------|--------|-------|---------|
| Number of tools | 3 | 1 | 67% fewer tools |
| Schema tokens | ~750 tokens | ~400 tokens | **~47% reduction** |

**Response Value (per API call):**
| Data Returned | Before | After |
|---------------|--------|-------|
| Filter columns | Yes | Yes |
| Select columns with metadata | No | Yes |
| Sortable columns | No | Yes |
| Default columns | No | Yes |
| Search columns | No | Yes |
| Column descriptions & types | No | Yes |

**Net Benefits:**
- **~350 fewer tokens** loaded per MCP session
- **5x richer** metadata per response
- LLMs can now discover valid `select_columns`, `order_column`, and `search` parameters
- Eliminates trial-and-error for column names

#### Changes

**Added:**
- `get_schema` tool in `superset/mcp_service/system/tool/get_schema.py`
- `ModelGetSchemaCore` in `mcp_core.py` for reusable schema discovery logic
- `schema_discovery.py` with column metadata definitions for all model types
- Comprehensive unit tests (6 new tests) with edge case coverage

**Removed:**
- `get_chart_available_filters.py` (deprecated)
- `get_dataset_available_filters.py` (deprecated)
- `get_dashboard_available_filters.py` (deprecated)
- Associated filter-only schemas from each module

**Updated:**
- `app.py` instructions to reference new `get_schema` tool
- MCP service CLAUDE.md documentation

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

**Before (3 separate tools, limited info):**
```json
// get_chart_available_filters response
{"filter_columns": {"slice_name": ["eq", "sw"], "viz_type": ["eq", "in"]}}
```

**After (1 unified tool, comprehensive info):**
```json
// get_schema response for model_type="chart"
{
  "schema_info": {
    "model_type": "chart",
    "select_columns": [
      {"name": "id", "description": "Chart ID", "type": "int", "is_default": true},
      {"name": "slice_name", "description": "Chart name", "type": "str", "is_default": true}
    ],
    "filter_columns": {"slice_name": ["eq", "sw"], "viz_type": ["eq", "in"]},
    "sortable_columns": ["id", "slice_name", "viz_type", "changed_on", "created_on"],
    "default_select": ["id", "slice_name", "viz_type", "uuid"],
    "default_sort": "changed_on",
    "default_sort_direction": "desc",
    "search_columns": ["slice_name", "description"]
  }
}
```

### TESTING INSTRUCTIONS

1. Start MCP server: `python -m superset.mcp_service.server`
2. Test via MCP client:
   ```python
   # Get chart schema
   result = await client.call_tool("get_schema", {"request": {"model_type": "chart"}})
   
   # Get dataset schema
   result = await client.call_tool("get_schema", {"request": {"model_type": "dataset"}})
   
   # Get dashboard schema  
   result = await client.call_tool("get_schema", {"request": {"model_type": "dashboard"}})
   ```
3. Run tests: `pytest tests/unit_tests/mcp_service/ -v`

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [x] Removes existing feature or API

___

## **CodeAnt-AI Description**
**Add unified get_schema tool and return minimal default columns for list APIs**

### What Changed
- Added a single get_schema tool that returns comprehensive schema metadata for charts, datasets, and dashboards (selectable columns with descriptions/types, filterable columns and operators, sortable columns, default select, default sort, and search columns).
- Removed the three legacy get_*_available_filters tools; callers should use get_schema instead.
- List endpoints (list_charts, list_datasets, list_dashboards) now apply field filtering by default: when select_columns is empty, only a minimal set of default columns (4 per model) is returned and heavy fields (e.g., columns, metrics, form_data, json_metadata) are excluded unless explicitly requested.
- List responses now include columns_available and sortable_columns metadata so clients can discover which columns can be requested or used for ordering.
- Added static schema resources (superset://schema/chart, /dataset, /dashboard, and /all) and updated default MCP instructions to recommend get_schema; tests updated/added to cover schema discovery and default-column behavior.

### Impact
 `✅ ~47% fewer tool tokens per MCP session`
 `✅ Smaller default list responses`
 `✅ Fewer trial-and-error column requests`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: superset/mcp_service/chart/tool/list_charts.py [hunk 1]
```diff
@@ -132,20 +133,16 @@ def _serialize_chart(
...
-            # Return dict with context - FastMCP will serialize it
-            return result.model_dump(
-                mode="json", context={"select_columns": request.select_columns}
-            )
-
-        # No filtering - return full result as dict
-        return result.model_dump(mode="json")
+        # Apply field filtering via serialization context
+        # Use requested columns or defaults - always filter to reduce token usage
+        columns_to_filter = request.select_columns or DEFAULT_CHART_COLUMNS
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/chart/tool/list_charts.py]: Addressed - Now using `result.columns_requested` which is already resolved/validated by ModelListCore instead of raw `request.select_columns`.

File: superset/mcp_service/dashboard/tool/list_dashboards.py [hunk 2]
```diff
@@ -108,17 +112,11 @@ def _serialize_dashboard(
...
-        # Return dict with context - FastMCP will serialize it
-        return result.model_dump(
-            mode="json", context={"select_columns": request.select_columns}
-        )
-
-    # No filtering - return full result as dict
-    return result.model_dump(mode="json")
+    # Apply field filtering via serialization context
+    # Use requested columns or defaults - always filter to reduce token usage
+    columns_to_filter = request.select_columns or DEFAULT_DASHBOARD_COLUMNS
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/dashboard/tool/list_dashboards.py]: Addressed - Same fix applied: now using `result.columns_requested` which is resolved by ModelListCore.

File: superset/mcp_service/dataset/tool/list_datasets.py [hunk 3]
```diff
@@ -144,20 +144,16 @@ def _serialize_dataset(
...
+        # Apply field filtering via serialization context
+        # Use requested columns or defaults - always filter to reduce token usage
+        columns_to_filter = request.select_columns or DEFAULT_DATASET_COLUMNS
+        await ctx.debug(
+            "Applying field filtering via serialization context: select_columns=%s"
+            % (columns_to_filter,)
+        )
+        return result.model_dump(
+            mode="json", context={"select_columns": columns_to_filter}
+        )
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/dataset/tool/list_datasets.py:160]: Addressed - Now returning `DatasetList.model_validate(filtered)` which properly converts the dict back to a Pydantic model instance.

File: superset/mcp_service/mcp_core.py [hunk 4]
```diff
@@ -106,6 +106,8 @@ def __init__(
         search_columns: List[str],
         list_field_name: str,
         output_list_schema: Type[L],
+        all_columns: List[str] | None = None,
+        sortable_columns: List[str] | None = None,
         logger: logging.Logger | None = None,
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/mcp_core.py]: Addressed - Moved `logger` parameter before `all_columns` and `sortable_columns` to preserve backwards compatibility with existing callers.

File: superset/mcp_service/mcp_core.py [hunk 5]
```diff
@@ -195,6 +199,8 @@ def get_keys(obj: BaseModel | dict[str, Any] | Any) -> List[str]:
             "has_next": page < total_pages - 1,
             "columns_requested": columns_requested,
             "columns_loaded": columns_to_load,
+            "columns_available": self.all_columns,
+            "sortable_columns": self.sortable_columns,
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/mcp_core.py:213]: Addressed - Now returning `list(...)` copies in the result to prevent external mutation of internal state.

File: superset/mcp_service/system/tool/get_schema.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    Args:
+        model_type: One of "chart", "dataset", or "dashboard"
+
+    Returns:
+        Comprehensive schema information for the requested model type
+    """
+    await ctx.info(f"Getting schema for model_type={request.model_type}")
+
+    # Get the appropriate core instance and run it
+    core = _SCHEMA_CORES[request.model_type]
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/system/tool/get_schema.py]: Addressed - Now using `.get()` with explicit error handling that returns a clear error message listing valid model types.

File: tests/unit_tests/mcp_service/dashboard/tool/test_dashboard_tools.py [hunk 7]
```diff
@@ -513,6 +514,96 @@ async def test_list_dashboards_custom_uuid_slug_columns(mock_list, mcp_server):
...
+        dashboard.certified_by = None
+        dashboard.certification_details = None
+        dashboard.json_metadata = None
+        dashboard.position_json = None
+        dashboard.is_managed_externally = False
+        dashboard.external_url = None
+        dashboard.thumbnail_url = None
+        dashboard.roles = []
+        dashboard.charts = []
+        mock_list.return_value = ([dashboard], 1)
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply tests/unit_tests/mcp_service/dashboard/tool/test_dashboard_tools.py:594]: The tests pass with the current implementation. The serialization uses `serialize_dashboard_object()` which doesn't rely on `_mapping` - it extracts fields directly from the object attributes.

File: superset/mcp_service/system/tool/get_schema.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+"""
+Unified schema discovery tool for MCP service.
+
+This tool consolidates schema discovery for all model types (chart, dataset,
+dashboard) into a single endpoint, reducing token usage and API calls.
+"""
+
+import logging
+from typing import Literal
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/system/tool/get_schema.py]: Not applicable - the implementation doesn't use `asyncio.to_thread`. The core classes are synchronous and the tool function just awaits the result directly.

File: superset/mcp_service/dataset/schemas.py [hunk 9]
```diff
@@ -200,6 +179,14 @@ class DatasetList(BaseModel):
     has_next: bool
     columns_requested: List[str] | None = None
     columns_loaded: List[str] | None = None
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/dataset/schemas.py]: Addressed - Updated all three schema files (chart, dashboard, dataset) to use `Field(default_factory=list)` for `columns_requested` and `columns_loaded`, ensuring consistent list defaults across all model types.

File: superset/mcp_service/mcp_core.py [hunk 10]
```diff
@@ -473,34 +479,96 @@ def _generate_instance_info(self) -> BaseModel:
...
 
-    def run_tool(self) -> S:
+    def _get_filter_columns(self) -> Dict[str, List[str]]:
+        """Get filterable columns and operators from the DAO."""
         try:
             filterable = self.dao_class.get_filterable_columns_and_operators()
-            # Ensure column_operators is a plain dict, not a custom type
-            column_operators = dict(filterable)
-            response = self.output_schema(column_operators=column_operators)
+            return dict(filterable)
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/mcp_core.py]: Addressed - Added defensive handling in `_get_filter_columns()`: checks for None return value, handles both dict and dict-like objects, and logs a warning with the unexpected type if conversion fails.

File: superset/mcp_service/system/resources/__init__.py [hunk 11]
```diff
@@ -18,4 +18,7 @@
 """System resources for Superset MCP service"""
 
 # Import to register resources when module is imported
-from . import instance_metadata  # noqa: F401
+from . import (
+    instance_metadata,  # noqa: F401
+    schema_discovery,  # noqa: F401
+)
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/system/resources/__init__.py:24]: Addressed - Changed from grouped tuple import to separate import lines for better clarity and per-import linting.

File: tests/unit_tests/mcp_service/chart/tool/test_list_charts.py [hunk 12]
```diff
@@ -170,3 +173,37 @@ def test_model_dump_serialization(self):
...
+class TestChartDefaultColumnFiltering:
+    """Test default column filtering behavior for charts."""
+
+    def test_minimal_default_columns_constant(self):
+        """Test that minimal default columns are properly defined."""
+        from superset.mcp_service.common.schema_discovery import CHART_DEFAULT_COLUMNS
+
+        # Should have exactly 4 minimal columns
+        assert len(CHART_DEFAULT_COLUMNS) == 4
+        assert set(CHART_DEFAULT_COLUMNS) == {"id", "slice_name", "viz_type", "uuid"}
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply tests/unit_tests/mcp_service/chart/tool/test_list_charts.py]: Addressed - Replaced the exact length assertion with `issubset` check: `required_columns.issubset(set(CHART_DEFAULT_COLUMNS))`. This ensures the required columns are present while allowing future additions.

File: tests/unit_tests/mcp_service/chart/tool/test_list_charts.py [hunk 13]
```diff
@@ -170,3 +173,37 @@ def test_model_dump_serialization(self):
...
+
+    def test_explicit_select_columns(self):
+        """Test that explicit select_columns can include non-default columns."""
+        request = ListChartsRequest(
+            select_columns=["id", "slice_name", "description", "form_data"]
+        )
+        assert "description" in request.select_columns
+        assert "form_data" in request.select_columns
+        # These are explicitly requested, not defaults
+        assert len(request.select_columns) == 4
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply tests/unit_tests/mcp_service/chart/tool/test_list_charts.py]: Addressed - Replaced the loose length assertion with a precise set equality check: `assert set(request.select_columns) == {"id", "slice_name", "description", "form_data"}`

File: superset/mcp_service/system/resources/schema_discovery.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        DATASET_SEARCH_COLUMNS,
+        DATASET_SELECT_COLUMNS,
+        DATASET_SORTABLE_COLUMNS,
+    )
+
+    schemas = {
+        "chart": {
+            "model_type": "chart",
+            "select_columns": [
+                {"name": col.name, "description": col.description, "type": col.type}
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/system/resources/schema_discovery.py]: Addressed - Added `_safe_col_dict()` helper function that uses `getattr()` with defaults and `str()` coercion to ensure JSON-safe serialization of column metadata.

File: superset/mcp_service/system/resources/schema_discovery.py [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+            "default_columns": CHART_DEFAULT_COLUMNS,
+            "sortable_columns": CHART_SORTABLE_COLUMNS,
+            "search_columns": CHART_SEARCH_COLUMNS,
+            "default_sort": "changed_on",
+            "default_sort_direction": "desc",
+        },
+        "dataset": {
+            "model_type": "dataset",
+            "select_columns": [
+                {"name": col.name, "description": col.description, "type": col.type}
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/system/resources/schema_discovery.py]: Addressed - Same fix applied using the shared `_safe_col_dict()` helper for JSON-safe serialization.

File: superset/mcp_service/system/resources/schema_discovery.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+            "default_columns": DATASET_DEFAULT_COLUMNS,
+            "sortable_columns": DATASET_SORTABLE_COLUMNS,
+            "search_columns": DATASET_SEARCH_COLUMNS,
+            "default_sort": "changed_on",
+            "default_sort_direction": "desc",
+        },
+        "dashboard": {
+            "model_type": "dashboard",
+            "select_columns": [
+                {"name": col.name, "description": col.description, "type": col.type}
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply superset/mcp_service/system/resources/schema_discovery.py]: Addressed - Same fix applied using the shared `_safe_col_dict()` helper for JSON-safe serialization.

File: tests/unit_tests/mcp_service/system/tool/test_get_schema.py [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+    ):
+        """Test that select_columns include metadata (description, type)."""
+        mock_filters.return_value = {}
+
+        async with Client(mcp_server) as client:
+            result = await client.call_tool(
+                "get_schema", {"request": {"model_type": "chart"}}
+            )
+
+            data = json.loads(result.content[0].text)
```
[the remark this answers is not in the document]
Comment (aminghadersohi) [reply tests/unit_tests/mcp_service/system/tool/test_get_schema.py:268]: Already addressed - All tests have `assert result.content is not None` before accessing `result.content[0].text`. This assertion exists in every test that accesses content.

File: superset/mcp_service/common/schema_discovery.py [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+
+    schema_info: ModelSchemaInfo = Field(
+        ..., description="Comprehensive schema information"
+    )
+
+
+# Column definitions for each model type
+# These are the source of truth for available columns
+
+CHART_SELECT_COLUMNS: List[ColumnMetadata] = [
```
Comment (michael-s-molina) [inline superset/mcp_service/common/schema_discovery.py]: @aminghadersohi The models already contain the columns, types, descriptions, etc. Do you think it's possible to obtain the schema dynamically from existing models instead of duplicating the definitions?

Comment (aminghadersohi) [reply superset/mcp_service/common/schema_discovery.py]: good point. let me do that and see what happens

Comment (aminghadersohi) [reply superset/mcp_service/common/schema_discovery.py]: done

File: superset/mcp_service/common/schema_discovery.py [hunk 19]
```diff
@@ -0,0 +1,422 @@
...
+
+"""
+Schema discovery models for MCP tools.
+
+These schemas provide comprehensive metadata about available columns,
+filters, and sorting options for each model type (chart, dataset, dashboard).
+Column metadata is extracted dynamically from SQLAlchemy models.
+"""
+
+from typing import Any, Dict, List, Literal, Type
```
Comment (villebro) [inline superset/mcp_service/common/schema_discovery.py]: Nit, we should use `dict`, `list`, `"my_literal"` instead of these: 
```suggestion
from typing import Any, Type
```

**Review by villebro — APPROVED** [review]
Great improvement! Minor nit, but other than that LGTM.

**Comment by sadpandajoe** [comment]
@dosubot what's the reason for the breaking change label?
